internal/appsec/remoteconfig.go

```diff
@@ -41,9 +43,89 @@ func statusesFromUpdate(u remoteconfig.ProductUpdate, ack bool, err error) map[s
 	return statuses
 }
 
-// asmFeaturesCallback deserializes an ASM_FEATURES configuration received through remote config
-// and starts/stops appsec accordingly. Used as a callback for the ASM_FEATURES remote config product.
-func (a *appsec) asmFeaturesCallback(u remoteconfig.ProductUpdate) map[string]rc.ApplyStatus {
+func mergeMaps[K comparable, V any](m1 map[K]V, m2 map[K]V) map[K]V {
+	merged := make(map[K]V)
+	for key, value := range m1 {
+		merged[key] = value
+	}
+	for key, value := range m2 {
+		merged[key] = value
+	}
+	return merged
+}
+
+// TODO: debug log + comments to explain
+func (a *appsec) asmUmbrellaCallback(updates map[string]remoteconfig.ProductUpdate) map[string]rc.ApplyStatus {
+	r := a.cfg.rulesManager.clone()
+	statuses := map[string]rc.ApplyStatus{}
+	// Process remote activation update separately
+	// This makes it easier to short circuit the following loop that focuses on rules related updates
+	if u, ok := updates[rc.ProductASMFeatures]; ok {
+		statuses = mergeMaps(statuses, a.handleASMFeatures(u))
+		delete(updates, rc.ProductASMFeatures)
+	}
+
+	// Process rules related updates
+	for p, u := range updates {
+		switch p {
+		case rc.ProductASMData:
+			rulesData, status := mergeRulesData(u)
+			statuses = mergeMaps(statuses, status)
+			r.addEdit("asmdata", rulesFragment{RulesData: rulesData})
+		case rc.ProductASMDD:
+			if len(u) > 1 { // Don't process configs if more than one is received for ASM_DD
+				log.Debug("appsec: Remote config: more than one config received for ASM_DD. Updates won't be applied")
+				statuses = mergeMaps(statuses, statusesFromUpdate(u, true, errors.New("More than one config received for ASM_DD")))
+				continue
+			}
+			for path, data := range u {
+				if data == nil {
+					r.changeBase(defaultRulesFragment(), "")
+					break
+				}
+				var newBase rulesFragment
+				if err := json.Unmarshal(data, &newBase); err != nil {
+					statuses[path] = genApplyStatus(true, err)
+					break
+				}
+				r.changeBase(newBase, path)
+				statuses[path] = genApplyStatus(true, nil)
+			}
+		case rc.ProductASM:
+			for path, data := range u {
+				statuses[path] = genApplyStatus(true, nil)
+				if data == nil {
+					r.removeEdit(path)
+					continue
+				}
+				var f rulesFragment
+				if err := json.Unmarshal(data, &f); err != nil || !f.validate() {
+					statuses[path] = genApplyStatus(true, err)
+				} else {
+					r.addEdit(path, f)
+				}
+			}
+		default:
+			log.Debug("appsec: Remote config: unsubscribed product %s. Ignoring", p)
+		}
+	}
+
+	r.compile()
+	data := r.raw()
+	log.Debug("appsec: Remote config: final compiled rulesManager: %v", data)
+	if err := a.swapWAF(data); err != nil {
+		for k := range statuses {
+			statuses[k] = genApplyStatus(true, err)
+		}
+	} else {
+		*a.cfg.rulesManager = *r
```

Comment:
Why a shallow copy here instead of a pointer replacement?
```suggestion
		a.cfg.rulesManager = r
```